pkg/rke2/control_plane.go

```diff
@@ -0,0 +1,343 @@
+}
+
+// HasDeletingMachine returns true if any machine in the control plane is in the process of being deleted.
+func (c *ControlPlane) HasDeletingMachine() bool {
+	return len(c.Machines.Filter(machinefilters.HasDeletionTimestamp)) > 0
+}
+
+// MachinesNeedingRollout return a list of machines that need to be rolled out.
+func (c *ControlPlane) MachinesNeedingRollout() FilterableMachineCollection {
+	// Ignore machines to be deleted.
+	machines := c.Machines.Filter(machinefilters.Not(machinefilters.HasDeletionTimestamp))
+
+	// Return machines if they are scheduled for rollout or if with an outdated configuration.
+	return machines.AnyFilter(
+		// Machines that are scheduled for rollout (RCP.Spec.UpgradeAfter set, the UpgradeAfter deadline is expired, and the machine was created before the deadline).
+		machinefilters.ShouldRolloutAfter(&c.reconciliationTime),
+		// Machines that do not match with RCP config.
+		machinefilters.Not(machinefilters.MatchesRCPConfiguration(c.infraResources, c.rke2Configs, c.RCP)),
+	)
+}
+
+// UpToDateMachines returns the machines that are up to date with the control
+// plane's configuration and therefore do not require rollout.
+func (c *ControlPlane) UpToDateMachines() FilterableMachineCollection {
+	return c.Machines.Difference(c.MachinesNeedingRollout())
+}
+
+// getInfraResources fetches the external infrastructure resource for each machine in the collection and returns a map of machine.Name -> infraResource.
+func getInfraResources(ctx context.Context, cl client.Client, machines FilterableMachineCollection) (map[string]*unstructured.Unstructured, error) {
+	result := map[string]*unstructured.Unstructured{}
+	for _, m := range machines {
+		infraObj, err := external.Get(ctx, cl, &m.Spec.InfrastructureRef, m.Namespace)
+		if err != nil {
+			if apierrors.IsNotFound(errors.Cause(err)) {
+				continue
+			}
+			return nil, errors.Wrapf(err, "failed to retrieve infra obj for machine %q", m.Name)
+		}
+		result[m.Name] = infraObj
+	}
+	return result, nil
+}
+
+// getRKE2Configs fetches the kubeadm config for each machine in the collection and returns a map of machine.Name -> RKE2Config.
+func getRKE2Configs(ctx context.Context, cl client.Client, machines FilterableMachineCollection) (map[string]*bootstrapv1.RKE2Config, error) {
+	result := map[string]*bootstrapv1.RKE2Config{}
+	for _, m := range machines {
+		bootstrapRef := m.Spec.Bootstrap.ConfigRef
+		if bootstrapRef == nil {
+			continue
+		}
+		machineConfig := &bootstrapv1.RKE2Config{}
+		if err := cl.Get(ctx, client.ObjectKey{Name: bootstrapRef.Name, Namespace: m.Namespace}, machineConfig); err != nil {
+			if apierrors.IsNotFound(errors.Cause(err)) {
+				continue
+			}
+			return nil, errors.Wrapf(err, "failed to retrieve bootstrap config for machine %q", m.Name)
+		}
+		result[m.Name] = machineConfig
+	}
+	return result, nil
+}
+
+// IsEtcdManaged returns true if the control plane relies on a managed etcd.
+func (c *ControlPlane) IsEtcdManaged() bool {
```

Comment:
are we going to support managed etcd?

Reply:
Not on RKE2 but it might be useful for K3S which does support it.